mersenne-31/src/x86_64_avx512/poseidon2.rs

```diff
@@ -1,58 +1,303 @@
-use p3_poseidon2::{matmul_internal, DiffusionPermutation};
-use p3_symmetric::Permutation;
+use alloc::vec::Vec;
+use core::arch::x86_64::{self, __m512i};
 
-use crate::{
-    DiffusionMatrixMersenne31, Mersenne31, PackedMersenne31AVX512,
-    POSEIDON2_INTERNAL_MATRIX_DIAG_16, POSEIDON2_INTERNAL_MATRIX_DIAG_24,
+use p3_poseidon2::{
+    mds_light_permutation, sum_15, sum_23, ExternalLayer, ExternalLayerConstants,
+    ExternalLayerConstructor, InternalLayer, InternalLayerConstructor, MDSMat4,
 };
 
-impl Permutation<[PackedMersenne31AVX512; 16]> for DiffusionMatrixMersenne31 {
-    fn permute_mut(&self, state: &mut [PackedMersenne31AVX512; 16]) {
-        matmul_internal::<Mersenne31, PackedMersenne31AVX512, 16>(
-            state,
-            POSEIDON2_INTERNAL_MATRIX_DIAG_16,
-        );
+use crate::{exp5, Mersenne31, PackedMersenne31AVX512, P, P_AVX512};
+
+/// The internal layers of the Poseidon2 permutation for Mersenne31.
+///
+/// This is optimized for the AVX512 architecture.
+#[derive(Debug, Clone)]
+pub struct Poseidon2InternalLayerMersenne31 {
+    pub(crate) internal_constants: Vec<Mersenne31>,
+    packed_internal_constants: Vec<__m512i>,
+}
+
+impl InternalLayerConstructor<PackedMersenne31AVX512> for Poseidon2InternalLayerMersenne31 {
+    /// We save the round constants in the {-P, ..., 0} representation instead of the standard
+    /// {0, ..., P} one. This saves several instructions later.
+    fn new_from_constants(internal_constants: Vec<Mersenne31>) -> Self {
+        Self::new_from_constants(internal_constants)
+    }
+}
+
+/// The external layers of the Poseidon2 permutation for Mersenne31.
+///
+/// This is optimized for the AVX512 architecture.
+#[derive(Clone)]
+pub struct Poseidon2ExternalLayerMersenne31<const WIDTH: usize> {
+    pub(crate) external_constants: ExternalLayerConstants<Mersenne31, WIDTH>,
+    packed_initial_external_constants: Vec<[__m512i; WIDTH]>,
+    packed_terminal_external_constants: Vec<[__m512i; WIDTH]>,
```

Comment:
Clarify that these are negative. Not everyone's going to look at where the object is constructed.